### ISSUE TITLE
Invalid Signature - Email Verification - HTTPS

### ISSUE THREAD

**Issue by aschbacd** [laravel/framework#28311]
- Laravel Version: 5.8.14
- PHP Version: 7.3.4
- Database Driver & Version: mysql 5.7

### Description:
When using HTTPS, the email validation link gets invalid and returns an HTTP Error with `403 Invalid signature.`. When using HTTP, everything works fine (but insecure).

**PS:** After enabling HTTPS every link uses HTTPS (also the one in the email).

### Steps To Reproduce:
1. Edit the file `app/Providers/AppServiceProvider.php` and add the following lines of code in the method `boot`:
```php
if(env('FORCE_HTTPS')) {
    URL::forceScheme('https');
}
```
2. Set the environment variable `FORCE_HTTPS` to `true`.
3. Run `php artisan config:cache` if the changes haven't been applied.

**Comment by 36864** [comment]
You're using the `env()` helper outside the configuration files and then caching your config. When you cache your config, `env()` no longer works. 

Can you try removing the condition and just setting `URL::forceScheme('https');`?

**Comment by laurencei** [comment]
Closing as @36864 has given the answer - dont use `env()` outside of config files.

**Comment by aschbacd** [comment]
I removed the `env()` condition and just set `URL::forceScheme('https');` but I still get the same error.

**Comment by aschbacd** [comment]
I found the issue with the help of @davidkroell. Before I was using container solutions to deploy my application (Docker and Heroku), but now I deployed it directly via Apache on a Debian system and the links are working fine (with HTTP and HTTPS).

I also didn't have to use `URL::forceScheme('https');` this time to get all links replaced with HTTPS ones.

**Comment by manfordbenjamin** [comment]
I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.

1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php

        if($this->app->environment('production'))
        {
            \URL::forceScheme('https');
        }

2. Allow all like below or configure proxies in https://laravel.com/docs/5.8/requests#configuring-trusted-proxies

     <?php

      namespace App\Http\Middleware;

      use Illuminate\Http\Request;
      use Fideloper\Proxy\TrustProxies as Middleware;

     class TrustProxies extends Middleware
     {
          /**
          * The trusted proxies for this application.
          *
          * @var array|string
          */
          protected $proxies = '*';

         /**
         * The headers that should be used to detect proxies.
         *
         * @var int
         */
         protected $headers = Request::HEADER_X_FORWARDED_ALL;
     }

**Comment by dhrumil4u360** [comment]
This issue is mainly introduced by adding following in your code,

`URL::forceScheme('https');`

If you are just trying to fix SSL URL links (When SSL is terminated in your load balancer & laravel server is receiving requests on port 80), you should not above approach. Instead, use approach given by Laravel. It will take care of exact this scenario.

https://laravel.com/docs/6.x/requests#configuring-trusted-proxies

If you have to have `forceScheme`, use approach mentioned by @manfordbenjamin

**Comment by narek-king** [comment]
In my case, it was the misconfiguration in NGINX
```
 location / {
    try_files $uri $uri/ /index.php?$query_string;
}
```
the `$query_string` was missing.

**Comment by za2n456** [comment]
> I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> 
> 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
>    ```
>     if($this->app->environment('production'))
>     {
>         \URL::forceScheme('https');
>     }
>    ```
> 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
>    ```
>    <?php
>    
>    namespace App\Http\Middleware;
>    
>    use Illuminate\Http\Request;
>    use Fideloper\Proxy\TrustProxies as Middleware;
>    
>    class TrustProxies extends Middleware
>    {
>       /**
>       * The trusted proxies for this application.
>       *
>       * @var array|string
>       */
>       protected $proxies = '*';
>    
>       /**
>       * The headers that should be used to detect proxies.
>       *
>       * @var int
>       */
>       protected $headers = Request::HEADER_X_FORWARDED_ALL;
>    }
>    ```

Thanks man it works.

**Comment by syedskspgon** [comment]
same issue faced..but i solved it by commenting few lines in "VerificationController"

public function __construct()
{
// $this->middleware('signed')->only('verify');
// $this->middleware('throttle:6,1')->only('verify', 'resend');
}

I hope this will works.
i think, it mainly disabling email expiry time

**Comment by supankar** [comment]
> > I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> > 
> > 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
> >    ```
> >     if($this->app->environment('production'))
> >     {
> >         \URL::forceScheme('https');
> >     }
> >    ```
> > 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
> >    ```
> >    <?php
> >    
> >    namespace App\Http\Middleware;
> >    
> >    use Illuminate\Http\Request;
> >    use Fideloper\Proxy\TrustProxies as Middleware;
> >    
> >    class TrustProxies extends Middleware
> >    {
> >       /**
> >       * The trusted proxies for this application.
> >       *
> >       * @var array|string
> >       */
> >       protected $proxies = '*';
> >    
> >       /**
> >       * The headers that should be used to detect proxies.
> >       *
> >       * @var int
> >       */
> >       protected $headers = Request::HEADER_X_FORWARDED_ALL;
> >    }
> >    ```
> 
> Thanks man it works.

yes, it works :)

**Comment by shashikanuwan** [comment]
What did you rename?

**Comment by shashikanuwan** [comment]
> Putting it out there for anyone else who might have encountered the same issue as me:
> 
> I too was running into this error. However, as it turned out, mine was caused by the fact that another route of my application already used the {hash} parameter. So when the email auth route was introduced (`email/verify/{id}/{hash}`) the route service provider tried to resolve a different kind of hash.
> 
> I renamed my application's hash parameter and it all worked.

What did you rename?

**Comment by shashikanuwan** [comment]
`403 | INVALID SIGNATURE`  Solve the problem 

first go to path, **_app->Http->Middleware->TrustProxies.php_**

change this code..
```
class TrustProxies extends Middleware
{
    protected $proxies = '*';

    protected $headers = Request::HEADER_X_FORWARDED_FOR | Request::HEADER_X_FORWARDED_HOST | Request::HEADER_X_FORWARDED_PORT | Request::HEADER_X_FORWARDED_PROTO;
}
```

**Comment by SoroushOracle** [comment]
If you use Nginx to web server, you must add following configuration to .conf file:

```
   location / {
         # remove it: try_files $uri $uri/ /index.php?q=$uri&$args;
         # replace this:
         try_files $uri $uri/ /index.php?$query_string;
   }
```

**Comment by ming1111dev** [comment]
Thanks @yawmanford 
That solved my problem.

**Comment by nguyenduytan** [comment]
> I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> 
> 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
>    ```
>     if($this->app->environment('production'))
>     {
>         \URL::forceScheme('https');
>     }
>    ```
> 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
>    ```
>    <?php
>    
>    namespace App\Http\Middleware;
>    
>    use Illuminate\Http\Request;
>    use Fideloper\Proxy\TrustProxies as Middleware;
>    
>    class TrustProxies extends Middleware
>    {
>       /**
>       * The trusted proxies for this application.
>       *
>       * @var array|string
>       */
>       protected $proxies = '*';
>    
>       /**
>       * The headers that should be used to detect proxies.
>       *
>       * @var int
>       */
>       protected $headers = Request::HEADER_X_FORWARDED_ALL;
>    }
>    ```

This work: protected $proxies = '*';

**Comment by Gunvantsinh** [comment]
> I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> 
> 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
>    ```
>     if($this->app->environment('production'))
>     {
>         \URL::forceScheme('https');
>     }
>    ```
> 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
>    ```
>    <?php
>    
>    namespace App\Http\Middleware;
>    
>    use Illuminate\Http\Request;
>    use Fideloper\Proxy\TrustProxies as Middleware;
>    
>    class TrustProxies extends Middleware
>    {
>       /**
>       * The trusted proxies for this application.
>       *
>       * @var array|string
>       */
>       protected $proxies = '*';
>    
>       /**
>       * The headers that should be used to detect proxies.
>       *
>       * @var int
>       */
>       protected $headers = Request::HEADER_X_FORWARDED_ALL;
>    }
>    ```

Thanks Man. It's working.

**Comment by seewhy17** [comment]
> > I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> > 
> > 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
> >    ```
> >     if($this->app->environment('production'))
> >     {
> >         \URL::forceScheme('https');
> >     }
> >    ```
> > 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
> >    ```
> >    <?php
> >    
> >    namespace App\Http\Middleware;
> >    
> >    use Illuminate\Http\Request;
> >    use Fideloper\Proxy\TrustProxies as Middleware;
> >    
> >    class TrustProxies extends Middleware
> >    {
> >       /**
> >       * The trusted proxies for this application.
> >       *
> >       * @var array|string
> >       */
> >       protected $proxies = '*';
> >    
> >       /**
> >       * The headers that should be used to detect proxies.
> >       *
> >       * @var int
> >       */
> >       protected $headers = Request::HEADER_X_FORWARDED_ALL;
> >    }
> >    ```
> 
> Thanks man it works.

Also works for me!!

**Comment by ettdro** [comment]
Also works for me, but I forgot to use the middleware in my App\Http\Kernel.php so it was never called. I just added it in `middlewareGroups` (that's where it goes for my case) and it works perfectly now. I messed around for a lot of time to simply realise I was actually not using it.

**Comment by hendisantika** [comment]
> I had a similar problem - it can be solved very easy if you use the TrustedProxy Middleware.
> 
> 1. Add \URL::forceScheme('https'); to the boot method in AppServiceProvider.php
>    ```
>     if($this->app->environment('production'))
>     {
>         \URL::forceScheme('https');
>     }
>    ```
> 2. Allow all like below or configure proxies as explained in the docs: https://laravel.com/docs/5.8/requests#configuring-trusted-proxies
>    ```
>    <?php
>    
>    namespace App\Http\Middleware;
>    
>    use Illuminate\Http\Request;
>    use Fideloper\Proxy\TrustProxies as Middleware;
>    
>    class TrustProxies extends Middleware
>    {
>       /**
>       * The trusted proxies for this application.
>       *
>       * @var array|string
>       */
>       protected $proxies = '*';
>    
>       /**
>       * The headers that should be used to detect proxies.
>       *
>       * @var int
>       */
>       protected $headers = Request::HEADER_X_FORWARDED_ALL;
>    }
>    ```

Also works for me!!

**Comment by bcag2** [comment]
in my Laravel v10 / breeze config, within a php/nginx docker container behind apache2 proxy, I replaced in  
_app/Providers/AppServiceProvider.php_:  
```php
    public function boot(): void
    {
        if($this->app->environment('production')) {
                        \URL::forceScheme('https');
                }
    }
// ----------------- by ----------->
    public function boot(): void
    {
		if (config('app.env') === 'production') {
			$this->app['request']->server->set('HTTPS', true);
		}
    }
```
and it works fine now !

**Comment by rizqyhi** [comment]
We've got the same issue on Laravel 11. Previously we setup `URL::forceHttps()` in `AppServiceProvider.php` to fix the URL generation problem. But then we migrated the server and put it behind Cloudflare. After that we start getting this invalid signature.

Then we set the `protected $proxies;` property into `protected $proxies = '*';` in `app/Http/Middleware/TrustProxies.php`. The problem is not going anywhere. After that we tried to remove the `URL::forceHttps()` hack. It is now working as expected, as @dhrumil4u360's [comment above](https://github.com/laravel/framework/issues/28311#issuecomment-559450853).
